content/posts/numpy/numpy-rng/index.md

```diff
@@ -0,0 +1,144 @@
+
+## Random number generation with NumPy
+
+When you import `numpy` in your python script, an RNG is created behind the scenes. This RNG is the one used when you generate a new random value using a function such as `np.random.random`. I will here refer to this RNG as the global NumPy RNG.
+
+Although not recommended, it is a common practice to reset the seed of this global RNG at the beginning of a script using the `np.random.seed` function. Fixing the seed at the beginning ensures that the script is reproducible: the same values and results will be produced each time you run it. However, although sometimes convenient, using the global NumPy RNG is a bad practice. A simple reason is that using global variables can lead to undesired side effects. For instance one might use `np.random.random` without knowing that the seed of the global RNG was set somewhere else in the codebase. Quoting [Numpy Enhancement Proposal (NEP) 19](https://numpy.org/neps/nep-0019-rng-policy.html) by Robert Kern:
+
+> The implicit global RandomState behind the `np.random.*` convenience functions can cause problems, especially when threads or other forms of concurrency are involved. Global state is always problematic. We categorically recommend avoiding using the convenience functions when reproducibility is involved. [...] The preferred best practice for getting reproducible pseudorandom numbers is to instantiate a generator object with a seed and pass it around.
+
+In short:
+
+- Instead of using `np.random.seed`, which reseeds the already created global NumPy RNG, and then using `np.random.*` functions, you should create a new RNG.
+- You should create an RNG at the beginning of your script (with your own seed if you want reproducibility) and use this RNG in the rest of the script.
+
+To create a new RNG you can use the `default_rng` function as illustrated in the [introduction of the random module documentation](https://numpy.org/doc/stable/reference/random/index.html):
+
+```python
+import numpy as np
+
+rng = np.random.default_rng()
+rng.random()  # generate a floating point number between 0 and 1
+```
+
+If you want to use a seed for reproducibility, [the NumPy documentation recommends](https://numpy.org/doc/stable/reference/random/index.html#quick-start) using a very large and unique number to have a different seed than anyone else. Another reason is that using a small number for the seed could lead to a bad first internal state of the RNG leading to a sequence of numbers that would not look like a random sequence, although `default_rng` now takes care of creating a good internal state from a small number. To obtain a good and unique seed you can rely on the [secrets module](https://docs.python.org/3/library/secrets.html). The following code will return a 128 bit number that you can then use as a seed:
+
+```python
+import secrets
+
+secrets.randbits(128)
```

Comment:
Given that `default_rng` does a good job of seeding from a small number, will 64 suffice? 32?

I suppose it does not matter whether you start with a small or large number, if all you want is reproducibility. Is the concern with small seeds not what happens when you instantiate multiple PRNGs, using x, x+1, x+2, etc?

Reply:
From my understanding using a large number prevents you (with higher probability) from using the same as someone else. So someone would be able to check if they can confirm the results of a study I did without using the same seed by chance. But please correct me if this not why we would want a different seed than anyone else?

The concern is the same for small and large seeds when you instantiate multiple RNGs using x, x+1 etc... you do not know if your seeds will lead to less independent numbers although again default_rng is supposed to do a good job at creating good initial states from close seeds.

Reply:
You are quite right. So it depends on whether you want independence from others (or subsequent runs of your own), or independence between streams. Often, we only want reproducibility.